test/src/test/java/jenkins/install/InstallUtilTest.java

```diff
@@ -0,0 +1,201 @@
+        // Reset the unit test flag back to its default.
+        Main.isUnitTest = true;
+    }
+
+    /**
+     * Test jenkins startup sequences and the changes to the startup type..
+     */
+    @Test
+    public void test_typeTransitions() {
+        // A new test instance
+        Assert.assertEquals(InstallState.NEW, InstallUtil.getInstallState());
+
+        // Save the last exec version. This will only be done by Jenkins after one of:
+        //   1. A successful run of the install wizard.
+        //   2. A success upgrade.
+        //   3. A successful restart.
+        InstallUtil.saveLastExecVersion();
+
+        // Fudge things a little now, pretending there's a restart...
+
+        // Now if we ask what is the InstallState, we should be told it's a RESTART because
+        // the install wizard is complete and the version matches the currently executing
+        // Jenkins version.
+        Assert.assertEquals(InstallState.RESTART, InstallUtil.getInstallState());
+
+        // Fudge things again, changing the stored version to something old, faking an upgrade...
+        InstallUtil.saveLastExecVersion("1.584");
+        Assert.assertEquals(InstallState.UPGRADE, InstallUtil.getInstallState());
+
+        // Fudge things yet again, changing the stored version to something very very new, faking a downgrade...
+        InstallUtil.saveLastExecVersion("1000.0");
+        Assert.assertEquals(InstallState.DOWNGRADE, InstallUtil.getInstallState());
+    }
+
+
+    /**
+     * Test jenkins startup sequences and the changes to the startup type..
+     */
+    @Test
+    public void test_getLastExecVersion() throws Exception {
+        // Delete the config file, forcing getLastExecVersion to return
+        // the default/unset version value.
+        InstallUtil.getConfigFile().delete();
+        Assert.assertEquals("1.0", InstallUtil.getLastExecVersion());
+
+        // Set the version to some stupid value and check again. This time,
+        // getLastExecVersion should read it from the file.
+        setStoredVersion("9.123");
+        Assert.assertEquals("9.123", InstallUtil.getLastExecVersion());
+    }
+
+    private void setStoredVersion(String version) throws Exception {
+        Field versionField = Jenkins.class.getDeclaredField("version");
+        versionField.setAccessible(true);
+        versionField.set(jenkinsRule.jenkins, version);
+        Assert.assertEquals(version, Jenkins.getStoredVersion().toString());
+        // Force a save of the config.xml
+        jenkinsRule.jenkins.save();
+    }
+
+    /**
+     * Validate proper statuses are persisted and install status is cleared when invoking appropriate methods on {@link InstallUtil}
+     */
+    @Test
+    public void testSaveAndRestoreInstallingPlugins() throws Exception {
+        final List<UpdateCenterJob> updates = new ArrayList<>();
+
+        final Map<String,String> nameMap = new HashMap<>();
+
+        new UpdateCenter() { // inner classes...
+		{
```

Comment:
this is multiple anonymous inner classes with defaul initializers and is not easy to grok...